content_script.js

```diff
@@ -1,82 +1,213 @@
-var filteredTags = ["script", "noscript", "style"];//, "link"];
+/**
+ * HTML serializer that takes a document and synchronously stores it as an array of strings and stores enough state to later
+ * asynchronously convert it into an html text file.
+ */
+class HTMLSerializer { // TODO: BECAUSE MULTIPLE CONTENT SCRIPTS CAN BE INJECTED, GIVES Identifier 'HTMLSerializer' has already been declared ERROR
+                       //       AND IF I ENCAPSULATE THE CLASS IN {} THE TESTS WONT BE ABLE TO INSTANTIATE AN INSTANCE
+  constructor() {
 
-function serializeTree(html, holes, element, depth) {
-  if (!element.tagName && element.nodeType != 3) {
-    // ignore elements that don't have tags and are not text
-  }
-  else if (element.tagName && filteredTags.indexOf(element.tagName.toLowerCase()) != -1) {
-    // filter out elements that are in filteredTags
-  }
-  else if (element.nodeType == 3) {
-    html.push(element.textContent);
+    /**
+     * @private {Set<string>} Contains the lower case tag names that should be ignored while serializing a document.
+     * @const
+     */
+    this.FILTERED_TAGS = new Set(['script', 'noscript', 'style', 'link']);
+
+    /**
+     * @public {Array<string>} This array represents the serialized html that makes up an element or document. 
+     */
+    this.html = [];
+
+    /**
+     * @public {Object<number, string>} The keys represent an index in this.html.  The value
+     *    is a url at which the resource that belongs at that index can be retrieved.  The resource will eventually
+     *    be converted to a data url.  Because any given document being serialized could be an iframe which is
+     *    nested 1 or more levels into the root document, the exact quotes that will be used to surround the data url
+     *    must be determined when the frames are being put together, so that they can be properly escaped.  As a
+     *    result, this.html has a filler index both immediately before, and immediately after where the data url will
+     *    be placed (3 filler indices in total).
+     */
+    this.srcHoles = {};
+
+    /**
+     * @public {Object<number, string>} The keys represent an index in this.html.  The value
```

Comment:
Nit "represent an index in |this.html|"  The | bracket emphasizes that you're referring to a variable name.